File: Week_03/46/permutations.go
package main

func permute(nums []int) [][]int {
	ans := make([][]int, 0)
	k := len(nums)
	visited := make(map[int]bool, k)
	option := make([]int, k)
	var dfs func(idx int)
	dfs = func(idx int) {
		for _, n := range nums {
			if visited[n] {
				continue
			}
			option[idx] = n
			visited[n] = true
			if idx == k-1 {
				ans = append(ans, append([]int{}, option...))
			} else {
				dfs(idx + 1)
			}
			visited[n] = false
		}
	}
	dfs(0)
	return ans
}

File: Week_03/46/permutations.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        k = len(nums)
        option = [0] * k
        visited = [0] * k

        def dfs(idx: int):
            for i in range(k):
                if visited[i] == 0:
                    option[idx] = nums[i]
                    visited[i] = 1
                    if idx == k - 1:
                        ans.append(option.copy())
                    else:
                        dfs(idx + 1)
                    visited[i] = 0

        dfs(0)
        return ans
